extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/NoArgsConstructorProcessor.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.quarkus.arc.deployment;
+
+import java.lang.reflect.Modifier;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.function.BiFunction;
+
+import javax.enterprise.context.NormalScope;
+import javax.inject.Inject;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.AnnotationTarget.Kind;
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.DotName;
+import org.jboss.logging.Logger;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+
+import io.quarkus.arc.processor.BeanDeploymentValidator;
+import io.quarkus.arc.processor.BuiltinScope;
+import io.quarkus.arc.processor.InjectionTargetInfo;
+import io.quarkus.arc.processor.InjectionTargetInfo.TargetKind;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
+import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+
+public class NoArgsConstructorProcessor {
+
+    private static final Logger LOGGER = Logger.getLogger(NoArgsConstructorProcessor.class);
+
+    // Copied from java.lang.Class
+    // https://github.com/wildfly/jandex/issues/60
+    private static final int ANNOTATION = 0x00002000;
+
+    @Inject
+    BeanArchiveIndexBuildItem beanArchiveIndex;
+
+    @Inject
+    CombinedIndexBuildItem combinedIndex;
+
+    @Inject
+    BuildProducer<BytecodeTransformerBuildItem> transformers;
+
+    @Inject
+    BuildProducer<BeanDeploymentValidatorBuildItem> validators;
+
+    @BuildStep
+    public void addMissingConstructors() throws Exception {
+        Set<ClassInfo> targetClasses = new HashSet<>();
+        Set<DotName> normalScopes = initNormalScopes();
+
+        for (DotName normalScope : normalScopes) {
+            collectTargetClasses(targetClasses, normalScope);
+        }
+        for (Iterator<ClassInfo> iterator = targetClasses.iterator(); iterator.hasNext();) {
+            ClassInfo targetClass = iterator.next();
+            if (targetClass.hasNoArgsConstructor()
+                    || targetClass.methods().stream().anyMatch(m -> m.name().equals("<init>") && m.parameters().isEmpty())) {
```

Comment:
Can you explain what this line does different than `targetClass.hasNoArgsConstructor()`? 
I initially thought it might be for checking inherited constructors but the javadoc of `targetClass.methods()` suggests that inherited methods are not returned here.

Reply:
`hasNoArgsConstructor()` doesn't have to be supported. Well, it should work for any Jandex 1.2+ index but I just wanted to be sure ;-)

Reply:
Got it thanks! 
I would suggest however that we either remove it or adapt the other usages of `hasNoArgsConstructor` as well (`BeanDeployment` and `ResteasyServerCommonProcessor` show up in the codebase).

Reply:
Actually, we require Jandex 2.1+ and re-index dependencies if necessary (https://github.com/quarkusio/quarkus/blob/master/core/deployment/src/main/java/io/quarkus/deployment/index/ApplicationArchiveBuildStep.java#L200-L202). So `hasNoArgsConstructor()` should be enough. I'll update the PR.

Reply:
Sounds good, thanks!